db/memtable_list.cc

```diff
@@ -269,10 +269,243 @@ void MemTableListVersion::TrimHistory(autovector<MemTable*>* to_delete) {
+    // ...                +-----------------+ |   +-----------------+
+    //                            |           |
+    //                            V           |
+    //                    +-----------------+ |   +-----------------+
+    // imm_lists[n2]:     | mems_list[1][0] | +-> | mems_list[1][1] | ...
+    //                    +-----------------+ |   +-----------------+
+    //                            |           |
+    //                            V           |
+    //                    +-----------------+ |
+    // imm_lists[n3]:     |mems_list[M-1][0]| |
+    // ...                +-----------------+ |
+    //                            |           |
+    //                            +-----------+
+    // imm_lists[N - 1]:
+    //
+    MemTable* mem = nullptr;
+    MemTable* head = nullptr;
+    for (const auto& mems : mems_list) {
+      if (!mems->empty()) {
+        if (head == nullptr) {
+          head = mems->front();
+        }
+        if (mem != nullptr) {
+          mem->next_ = mems->front();
+        }
+        mem = mems->front();
+      }
+    }
+    if (mem != nullptr) {
+      mem->next_ = head;
+    }
+  }
+
+  int num = static_cast<int>(imm_lists.size());
+  assert(num == static_cast<int>(cfds.size()));
+  for (int k = 0; k != num; ++k) {
+    for (int i = 0; i != static_cast<int>(mems_list[k]->size()); ++i) {
+      assert(i == 0 || (*mems_list[k])[i]->GetEdits()->NumEntries() == 0);
+      (*mems_list[k])[i]->flush_completed_ = true;
+      (*mems_list[k])[i]->file_number_ = file_meta[k].fd.GetNumber();
+    }
+  }
+
+  Status s;
+  if (atomic_flush_commit_in_progress_) {
+    return s;
+  }
+
+  atomic_flush_commit_in_progress_ = true;
+
+  while (s.ok()) {
+    std::unordered_map<MemTable*, int> mem_to_idx;
+    autovector<int> next;
+    for (int i = 0; i != num; ++i) {
+      const std::list<MemTable*>& memlist = imm_lists[i]->current_->memlist_;
+      if (!memlist.empty()) {
+        const auto it = memlist.rbegin();
+        if ((*it)->flush_completed_) {
+          mem_to_idx.insert({*it, i});
+        }
+      }
+      next.push_back(-1);
+    }
+    if (mem_to_idx.empty()) {
+      // Nothing to commit
+      break;
+    }
+    for (int i = 0; i != num; ++i) {
+      const std::list<MemTable*>& memlist = imm_lists[i]->current_->memlist_;
+      if (!memlist.empty()) {
+        const auto it = memlist.rbegin();
+        assert((*it)->next_ != nullptr);
+        if ((*it)->flush_completed_) {
+          const auto next_it = mem_to_idx.find((*it)->next_);
+          if (next_it != mem_to_idx.end()) {
+            next[i] = next_it->second;
+          }
+        }
+      }
+    }
+    while (true) {
+      // Try to find a linked list formed by memtables at the rbegin side of
+      // each imm_lists. If such a linked list exist, we have found a new group
+      // to commit to MANIFEST.
+      int k = 0;
+      while (k < num && next[k] < 0) {
```

Comment:
What does it mean if `k` ends up nonzero but less than `num` after this loop? Is it an error, or is this setting up for a future PR to atomically flush a subset of CFs? (side note: if we are setting up for different flushes covering different number of CFs, I'm not sure this code handles single-CF flush).

Reply:
The `next` array is used for discovering the next atomic flush to commit. We commit the memtables in `imm_lists` in chronological order while considering the boundaries among different atomic flushes. Therefore, at the beginning of each iteration, we need to determine which memtables in which memtable lists to commit. The rules are the following:

* Within each memtable list, we always commit memtables with smaller memtable IDs because they are older.
* For each atomic flush spanning across multiple column families, only when ALL of its memtables are the oldest within the corresponding memtable list can we commit them.

`next[l] == m ( 0 <= m <= N)` means that the memtables in imm_lists[l] and imm_lists[m] belong to the same atomic flush, and they are both ready to commit (oldest in their corresponding imm_lists).

Reply:
Regarding your side note.
This code does handle single CF case as demonstrated by `MemTableListTest::FlushMultipleCFsTest` when `k == 2`.